File: serde-generate/runtime/golang/serde/binary_serializer.go
// Copyright (c) Facebook, Inc. and its affiliates
// SPDX-License-Identifier: MIT OR Apache-2.0

package serde

import (
	"bytes"
	"errors"
)

// `BinarySerializer` is a partial implementation of the `Serializer` interface.
// It is used as an embedded struct by the Bincode and LCS serializers.
type BinarySerializer struct {
	Buffer bytes.Buffer
}

func NewBinarySerializer() *BinarySerializer {
	return new(BinarySerializer)
}

// `serializeLen` to be provided by the extending struct.
func (s *BinarySerializer) SerializeBytes(value []byte, serializeLen func(uint64) error) error {
	serializeLen(uint64(len(value)))
	s.Buffer.Write(value)
	return nil
}

// `serializeLen` to be provided by the extending struct.
func (s *BinarySerializer) SerializeStr(value string, serializeLen func(uint64) error) error {
	return s.SerializeBytes([]byte(value), serializeLen)
}

func (s *BinarySerializer) SerializeBool(value bool) error {
	if value {
		return s.Buffer.WriteByte(1)
	}
	return s.Buffer.WriteByte(0)
}

func (s *BinarySerializer) SerializeUnit(value struct{}) error {
	return nil
}

// SerializeChar is unimplemented.
func (s *BinarySerializer) SerializeChar(value rune) error {
	return errors.New("unimplemented")
}

// SerializeF32 is unimplemented
func (s *BinarySerializer) SerializeF32(value float32) error {
	return errors.New("unimplemented")
}

// SerializeF64 is unimplemented
func (s *BinarySerializer) SerializeF64(value float64) error {
	return errors.New("unimplemented")
}

func (s *BinarySerializer) SerializeU8(value uint8) error {
	s.Buffer.WriteByte(byte(value))
	return nil
}

func (s *BinarySerializer) SerializeU16(value uint16) error {
	s.Buffer.WriteByte(byte(value))
	s.Buffer.WriteByte(byte(value >> 8))
	return nil
}

func (s *BinarySerializer) SerializeU32(value uint32) error {
	s.Buffer.WriteByte(byte(value))
	s.Buffer.WriteByte(byte(value >> 8))
	s.Buffer.WriteByte(byte(value >> 16))
	s.Buffer.WriteByte(byte(value >> 24))
	return nil
}

func (s *BinarySerializer) SerializeU64(value uint64) error {
	s.Buffer.WriteByte(byte(value))
	s.Buffer.WriteByte(byte(value >> 8))
	s.Buffer.WriteByte(byte(value >> 16))
	s.Buffer.WriteByte(byte(value >> 24))
	s.Buffer.WriteByte(byte(value >> 32))
	s.Buffer.WriteByte(byte(value >> 40))
	s.Buffer.WriteByte(byte(value >> 48))
	s.Buffer.WriteByte(byte(value >> 56))
	return nil
}

func (s *BinarySerializer) SerializeU128(value Uint128) error {
	s.SerializeU64(value.Low)
	s.SerializeU64(value.High)
	return nil
}

func (s *BinarySerializer) SerializeI8(value int8) error {
	s.SerializeU8(uint8(value))
	return nil
}

func (s *BinarySerializer) SerializeI16(value int16) error {
	s.SerializeU16(uint16(value))
	return nil
}

func (s *BinarySerializer) SerializeI32(value int32) error {
	s.SerializeU32(uint32(value))
	return nil
}

func (s *BinarySerializer) SerializeI64(value int64) error {
	s.SerializeU64(uint64(value))
	return nil
}

func (s *BinarySerializer) SerializeI128(value Int128) error {
	s.SerializeU64(value.Low)
	s.SerializeI64(value.High)
	return nil
}

func (s *BinarySerializer) SerializeOptionTag(value bool) error {
	return s.SerializeBool(value)
}

func (s *BinarySerializer) GetBufferOffset() uint64 {
	return uint64(s.Buffer.Len())
}

func (s *BinarySerializer) GetBytes() []byte {
	return s.Buffer.Bytes()
}

File: serde-generate/runtime/swift/Sources/Serde/BinarySerializer.swift
//  Copyright (c) Facebook, Inc. and its affiliates.

import Foundation

public enum BinarySerializerError: Error {
    case invalidValue(issue: String)
}

public class BinarySerializer: Serializer {
    public var output = OutputStream.toMemory()
    private var containerDepthBudget: Int

    public init(maxContainerDepth: Int) {
        output.open()
        containerDepthBudget = maxContainerDepth
    }

    deinit {
        output.close()
    }

    public func increase_container_depth() throws {
        if containerDepthBudget == 0 {
            throw BinarySerializerError.invalidValue(issue: "Exceeded maximum container depth")
        }
        containerDepthBudget -= 1
    }

    public func decrease_container_depth() {
        containerDepthBudget += 1
    }

    public func serialize_char(value _: Character) throws {
        throw BinarySerializerError.invalidValue(issue: "Not implemented: char serialization")
    }

    public func serialize_f32(value: Float) throws {
        try serialize_u32(value: value.bitPattern)
    }

    public func serialize_f64(value: Double) throws {
        try serialize_u64(value: value.bitPattern)
    }

    public func get_bytes() -> [UInt8] {
        return output.getBuffer()
    }

    public func serialize_str(value: String) throws {
        try serialize_bytes(value: Array(value.utf8))
    }

    public func serialize_bytes(value: [UInt8]) throws {
        try serialize_len(value: value.count)
        output.write(data: Data(value))
    }

    public func serialize_bool(value: Bool) throws {
        try writeByte(value ? 1 : 0)
    }

    public func serialize_unit(value _: Unit) throws {}

    func writeByte(_ value: UInt8) throws {
        if output.write(data: Data([value])) != 1 {
            throw BinarySerializerError.invalidValue(issue: "Error while outputting byte")
        }
    }

    public func serialize_u8(value: UInt8) throws {
        try writeByte(value)
    }

    public func serialize_u16(value: UInt16) throws {
        try writeByte(UInt8(truncatingIfNeeded: value))
        try writeByte(UInt8(truncatingIfNeeded: value >> 8))
    }

    public func serialize_u32(value: UInt32) throws {
        try writeByte(UInt8(truncatingIfNeeded: value))
        try writeByte(UInt8(truncatingIfNeeded: value >> 8))
        try writeByte(UInt8(truncatingIfNeeded: value >> 16))
        try writeByte(UInt8(truncatingIfNeeded: value >> 24))
    }

    public func serialize_u64(value: UInt64) throws {
        try writeByte(UInt8(truncatingIfNeeded: value))
        try writeByte(UInt8(truncatingIfNeeded: value >> 8))
        try writeByte(UInt8(truncatingIfNeeded: value >> 16))
        try writeByte(UInt8(truncatingIfNeeded: value >> 24))
        try writeByte(UInt8(truncatingIfNeeded: value >> 32))
        try writeByte(UInt8(truncatingIfNeeded: value >> 40))
        try writeByte(UInt8(truncatingIfNeeded: value >> 48))
        try writeByte(UInt8(truncatingIfNeeded: value >> 56))
    }

    public func serialize_u128(value: UInt128) throws {
        try serialize_u64(value: value.low)
        try serialize_u64(value: value.high)
    }

    public func serialize_i8(value: Int8) throws {
        try serialize_u8(value: UInt8(bitPattern: value))
    }

    public func serialize_i16(value: Int16) throws {
        try serialize_u16(value: UInt16(bitPattern: value))
    }

    public func serialize_i32(value: Int32) throws {
        try serialize_u32(value: UInt32(bitPattern: value))
    }

    public func serialize_i64(value: Int64) throws {
        try serialize_u64(value: UInt64(bitPattern: value))
    }

    public func serialize_i128(value: Int128) throws {
        try serialize_u64(value: value.low)
        try serialize_i64(value: value.high)
    }

    public func serialize_option_tag(value: Bool) throws {
        try writeByte(value ? 1 : 0)
    }

    public func get_buffer_offset() -> Int {
        return output.getOffset()
    }

    public func serialize_len(value _: Int) throws {
        assertionFailure("Not implemented")
    }

    public func serialize_variant_index(value _: UInt32) throws {
        assertionFailure("Not implemented")
    }

    public func sort_map_entries(offsets _: [Int]) {
        assertionFailure("Not implemented")
    }
}
